### PR TITLE
Parse unsafe attributes

### PR DESCRIPTION
**Pull request by carbotaniuman**
Initial parse implementation for #123757

This is the initial work to parse unsafe attributes, which is represented as an extra `unsafety` field in `MetaItem` and `AttrItem`. There's two areas in the code where it appears that parsing is done manually and not using the parser stuff, and I'm not sure how I'm supposed to thread the change there.

### CODE REVIEW

File: compiler/rustc_feature/src/builtin_attrs.rs
Comment (michaelwoerister) [inline compiler/rustc_feature/src/builtin_attrs.rs:197]: Maybe move the `unsafe` variant before the regular ones so that they are matched first. That should make it less likely that there is an accidental match where `unsafe` is matched as `$attr`.

File: compiler/rustc_feature/src/unstable.rs
Comment (michaelwoerister) [inline compiler/rustc_feature/src/unstable.rs:631]: The doc comment needs an update too.

**Comment by michaelwoerister** [comment]
Thanks for the PR, @carbotaniuman!

Would you mind also adding a test case where using `unsafe(..)` works as expected and one where it is used in a nested context, like `#[cfg_attr(foo, unsafe(no_mangle)]`?

**Comment by carbotaniuman** [comment]
I've added stuff to handle the `$meta` matcher in `macro_rules!`. I'm not sure if it's needed, but it's technically breaking, and RFC 3531 suggests I do.

File: compiler/rustc_ast/src/token.rs
Comment (michaelwoerister) [inline compiler/rustc_ast/src/token.rs:915]: Is this actually needed? The grammar doesn't change, right?

Comment (carbotaniuman) [reply compiler/rustc_ast/src/token.rs:915]: I'm not sure...

https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=5ccbe0db996ad629df3df4dc2119e2cd
```rust
#[cfg_attr(not(debug), unsafe(Foo))]
struct Foo;
```

The above is a parse error right now as it expects a path and not a keyword. There were 2 implementations strategies I considered, one of which was defining a macro `unsafe` that created the `Attribute` with the proper unsafety, but given that I would still need syntax changes (to deal with the `unsafe`), I opted to instead change the core `AttrItem` itself to have an `unsafety` field. So I'm not really sure if a) this is the right way of doing things and b) if I need to make the macro changes.

File: compiler/rustc_parse/src/parser/attr.rs
Comment (michaelwoerister) [inline compiler/rustc_parse/src/parser/attr.rs:274]: I expect this to result in a slightly weird error message when doing something like `#[unsafe(unsafe(no_mangle))]`. Can we add a test case for that?

File: compiler/rustc_parse/src/parser/nonterminal.rs
Comment (michaelwoerister) [inline compiler/rustc_parse/src/parser/nonterminal.rs:67]: Isn't the `token.is_keyword(kw::Unsafe)` case covered by the first match arm via `token::Ident(..)`? I don't see a separate `TokenKind` for keywords.

File: compiler/rustc_parse/src/parser/nonterminal.rs
Comment (michaelwoerister) [inline compiler/rustc_parse/src/parser/nonterminal.rs:178]: Again, I'm not sure this is necessary because using `unsafe(...)` currently fails to parse (see [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c728c2e4abf32a38d67ce89ab665b3f9)). So I think this should fall under the following paragraph from [RFC 3531](https://rust-lang.github.io/rfcs/3531-macro-fragment-policy.html):

> In cases where we’re adding new syntax and updating the grammar to include that new syntax, if we can update the corresponding fragment specifier simultaneously to match the new grammar in such a way that we do not risk changing the behavior of existing macros (i.e., because the new syntax previously would have failed parsing), then we will do that so as to prevent or minimize divergence between the fragment specifier and the new grammar. 

I.e. the question is: Can the addition of `unsafe(..)` change the behavior of any currently valid macros / macro invocations?

**Comment by michaelwoerister** [comment]
I think we need some expert input. There are two main questions:

- The PR's current implementation strategy for [unsafe attributes](https://github.com/rust-lang/rfcs/pull/3325) is to add an `unsafety` field to `AttrItem` and `MetaItem` and handled that during parsing. I.e. `#[unsafe(Something)]` will end up as a flat `Something` `AttrItem` as opposed to an `unsafe` `AttrItem` with `Something` nested as an argument. Given that the RFC says that unsafe attributes cannot be nested or complex, that implementation strategy makes sense to me. But I'm no parsing/macros expert so it would be good to get additional opinions here.

- The PR currently introduces a new `meta_2021` fragment specifier to account for the unsafe attributes changes (see [RFC 3531 - Macro matcher fragment specifiers edition policy](https://rust-lang.github.io/rfcs/3531-macro-fragment-policy.html)). However, I think that is not necessary since `#[unsafe(something)]` does not currently parse (see [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c728c2e4abf32a38d67ce89ab665b3f9)). Therefore starting to allow it for the existing `meta` fragment specifier should not be a breaking change and does not need to happen on an edition boundary. Does this reasoning overlook something?

@rustbot ping parser

**Comment by michaelwoerister** [comment]
It looks like rustbot does not allow pinging [adhoc groups](https://github.com/rust-lang/rust/blob/fcc06c894b17f4d0c80b8934ea5f27faa894c960/triagebot.toml#L908-L915), so pinging manually regarding [the questions in the comment above](https://github.com/rust-lang/rust/pull/124214#issuecomment-2090195007): @compiler-errors @davidtwco @estebank @nnethercote @petrochenkov @spastorino

**Comment by petrochenkov** [comment]
>The PR's current implementation strategy for [unsafe attributes](https://github.com/rust-lang/rfcs/pull/3325) is to add an `unsafety` field to `AttrItem` and `MetaItem` and handled that during parsing. I.e. `#[unsafe(Something)]` will end up as a flat `Something` `AttrItem` as opposed to an `unsafe` `AttrItem` with `Something` nested as an argument.

This is the right strategy.
It would be more clear if the syntax was different, like `#[unsafe my_attr]`.
The confusing parenthesized syntax was selected in the end, but it doesn't change the semantics.

**Comment by petrochenkov** [comment]
>Macro matcher fragment specifiers edition policy](https://rust-lang.github.io/rfcs/3531-macro-fragment-policy.html)). However, I think that is not necessary since `#[unsafe(something)]` does not currently parse (see [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c728c2e4abf32a38d67ce89ab665b3f9)). 

It would be good if `unsafe` could fit into the existing `meta` matcher.

There may be some issues with ambiguities between macro arms in cases similar to
```rust
macro_rules! m {
    (unsafe) => { 1 }
    ($meta:meta) => { 2 }
}

m!(unsafe(no_mangle));
```
This needs some testing.

**Comment by michaelwoerister** [comment]
> This is the right strategy.

Thanks for the feedback! 

> It would be more clear if the syntax was different, like #[unsafe my_attr].
> The confusing parenthesized syntax was selected in the end, but it doesn't change the semantics.

Yes, I read through the RFC discussion thread today. I'm going to stay neutral regarding the choice 🙂
From a grammar point of view `#[unsafe my_attr]` would indeed be clearer.

**Comment by michaelwoerister** [comment]
> It would be good if `unsafe` could fit into the existing `meta` matcher.

Let's try to make that work then.
 
> There may be some issues with ambiguities between macro arms in cases similar to
> 
> ```rust
> macro_rules! m {
>     (unsafe) => { 1 }
>     ($meta:meta) => { 2 }
> }
> 
> m!(unsafe(no_mangle));
> ```
> 
> This needs some testing.

To clarify, the above test case would be an example where the current and the new `meta` matcher could give different results?

Do you mean a case like the following where currently `:meta` does not match `unsafe(no_mangle)`, so we get the second arm, but with the new behavior we would get the first one ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dcc18c4ed43c09e458412fafba173f98))?:

```rust
fn main() {
    macro_rules! foo {
        ($l:meta) => { 1 };
        (unsafe($x:ident)) => { 2 };
    }

    let x = foo!(unsafe(no_mangle));
    println!("{x}");
}
```

File: compiler/rustc_parse/src/parser/attr.rs
Comment (carbotaniuman) [reply compiler/rustc_parse/src/parser/attr.rs:274]:
```
error: expected identifier, found keyword `unsafe`
  --> /home/quydx/code/rust/tests/ui/attributes/unsafe/double-unsafe-in-attribute.rs:1:10
   |
LL | #[unsafe(unsafe(no_mangle))]
   |          ^^^^^^ expected identifier, found keyword
   |
help: escape `unsafe` to use it as an identifier
   |
LL | #[unsafe(r#unsafe(no_mangle))]
   |          ++

error: cannot find attribute `r#unsafe` in this scope
  --> /home/quydx/code/rust/tests/ui/attributes/unsafe/double-unsafe-in-attribute.rs:1:10
   |
LL | #[unsafe(unsafe(no_mangle))]
   |          ^^^^^^
```
The error is basically just what we have now. I can probably add a note saying that `unsafe(...)` doesn't nest.

**Comment by carbotaniuman** [comment]
Yeah, that `foo` macro is the main one I'm worried about. I think you could get away with it with some sort of migration lint, but I'm really not sure.

**Comment by michaelwoerister** [comment]
@carbotaniuman, can you bring this PR in a state where we don't introduce a separate `meta_2021` matcher and just modify the existing `meta` matcher? Then we can do a crater run to see what the impact is.

@rustbot author

**Comment by carbotaniuman** [comment]
Alright, I've reverted those commits.

@rustbot ready

**Comment by michaelwoerister** [comment]
Thanks, @carbotaniuman!

Let's do a try build for the crater run.
@bors try

I guess we'll need a build-and-test run since the difference in parsing does not necessarily result in a compile time error.

**Comment by michaelwoerister** [comment]
A code search on github does not yield anything that looks problematic:
[`/#\[unsafe\(.*\)\]/ language:Rust`](https://github.com/search?q=%2F%23%5C%5Bunsafe%5C%28.*%5C%29%5C%5D%2F+language%3ARust+&type=code)
[`/#\[cfg_attr\(.*,.*unsafe\(.*\)\)\]/ language:Rust`](https://github.com/search?q=%2F%23%5C%5Bcfg_attr%5C%28.*%2C.*unsafe%5C%28.*%5C%29%5C%29%5C%5D%2F+language%3ARust+&type=code)

**Comment by michaelwoerister** [comment]
@craterbot run mode=build-and-test

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Let's nominate to discuss and confirm that the only attributes that need to be marked as unsafe are:

- `no_mangle`
- `link_section`
- `export_name`

There's been discussion about `link` and `link_ordinal` also.

cc @RalfJung

**Comment by RalfJung** [comment]
[This](https://github.com/rust-lang/rust/issues/82499) is the best list I know. But I honestly don't have a good overview of which attributes we even have for controlling linker behavior, let alone which of them are unsafe.

Cc @bjorn3

**Comment by carbotaniuman** [comment]
I went and looked through the list of stable and unstable features, and I have collected the following. Note that I don't understand all of these features and probably missed some.

Stable:
- link? (may allowing pulling in libraries with ctors/dtors)
- no_mangle
- link_section
- export_name

Of note is that I don't think `link_ordinal` *needs* to be unsafe, as it's no different from just misspelling the name of a function to link to (more realistically getting it mixed up with something else), and the safety justification for that is in the unsafe extern RFC.

Unstable:
- ffi_pure
- ffi_const
- naked?

I am ignoring the internal `rustc_` attributes as they're perma unstable right now.

**Comment by RalfJung** [comment]
> I am ignoring the internal rustc_ attributes as they're perma unstable right now.

Would be good to audit these eventually, but not for the first PR, agreed.

> Of note is that I don't think link_ordinal needs to be unsafe, as it's no different from just misspelling the name of a function to link to (more realistically getting it mixed up with something else), and the safety justification for that is in the unsafe extern RFC.

Agreed, link_ordinal seems to be "the same" as link_name.

**Comment by carbotaniuman** [comment]
Looking at the crater results, it appears like there's no regressions from this change.

**Comment by michaelwoerister** [comment]
> Looking at the crater results, it appears like there's no regressions from this change.

I agree. I can't find anything that looks related this PR.

File: compiler/rustc_ast/src/attr/mod.rs
Comment (michaelwoerister) [inline compiler/rustc_ast/src/attr/mod.rs:370]: Can you expand this comment to give more information on what needs to be fixed?

**Review by michaelwoerister — COMMENTED** [review]
The PR looks good to merge now. Please expand the comment mentioned below. After that I'm happy to approve.

**Comment by michaelwoerister** [comment]
Actually, @carbotaniuman, let's also make #[ffi_const] and #[ffi_pure] unsafe too. Those definitely fit the bill. Not sure about #[naked] yet.

**Comment by michaelwoerister** [comment]
@traviscross, let me know if you'd like me to hold off on approving the PR until the lang-team has reviewed the set of attributes.

**Comment by RalfJung** [comment]
`#[naked]` seems very unsafe, doesn't it? Cc @Amanieu

**Comment by michaelwoerister** [comment]
> `#[naked]` seems very unsafe, doesn't it? Cc @Amanieu

I agree but it might be a similar case to `#[link_name]` because either the naked function itself has to `unsafe` or its entire body must be wrapped inside an unsafe block. But maybe that's a kind of unsafety of a different category?

**Comment by pnkfelix** [comment]
> @traviscross, let me know if you'd like me to hold off on approving the PR until the lang-team has reviewed the set of attributes.

I'm doing that review now, thanks @michaelwoerister

File: compiler/rustc_ast/src/attr/mod.rs
Comment (carbotaniuman) [reply compiler/rustc_ast/src/attr/mod.rs]: I've put a proper `FIXME` here. I expect I'll tackle this in a subsequent PR, so I'd prefer not having to block on this.

Comment (michaelwoerister) [reply compiler/rustc_ast/src/attr/mod.rs]: Thanks, yes that works me.

**Comment by carbotaniuman** [comment]
Is this blocked on the lang list? I'd like to get started on the migration + warning lint for this.

**Comment by traviscross** [comment]
> Is this blocked on the lang list? I'd like to get started on the migration + warning lint for this.

We'll check in on this on the lang call today and try to get this unblocked.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We didn't get to this in the meeting today, but we can unblock it regardless.  I talked with @pnkfelix.  He did mean for his comment to unblock it.  The final list is:

- `export_name`
- `link_section`
- `no_mangle`
- `ffi_const`
- `ffi_pure`

This PR puts everything under a nightly feature flag, so lang will get a final chance to review this in the stabilization PR.

@michaelwoerister / @carbotaniuman: This PR is OK to move forward in terms of lang.

File: compiler/rustc_feature/src/builtin_attrs.rs
Comment (michaelwoerister) [inline compiler/rustc_feature/src/builtin_attrs.rs:63]: Is `#[allow(dead_code)]` still needed?

**Review by michaelwoerister — APPROVED** [review]
This is r=me after rebasing and looking into the nit below.
Thanks for all the work you put into this, @carbotaniuman!

**Comment by michaelwoerister** [comment]
Thanks for the update, @traviscross!

**Pull request merged**
